### PR TITLE
Unit tests for losses.

### PR DESCRIPTION
**Pull request by yhtang**
In addition to tests, I also relocated the `reduction` argument of the loss functions from `__call__` to `__init__`. `ff.factorize` was also updated to accept either class-valued or object-valued loss arguments.

### CODE REVIEW

File: docs/pages/user-guide/factorize.md [hunk 1]
```diff
@@ -107,8 +107,7 @@ a custom optimizer:
 
 The loss function used in the optimization process can be changed as well as its
 (hyper) parameters. This adheres to the FunFact [Loss API](../../../api/loss/).
-All keyword arguments that are input to `factorize` are passed to the loss
-function.
+The `optimizer_options` dictionary argument to the `factorize` will be forwarded as keyword arguments to initialize the loss function.
```
[the remark this answers is not in the document]
Comment (yhtang) [reply docs/pages/user-guide/factorize.md]: Right. I was probably drunk-coding then. It is now fixed (eliminated via the class-or-callable idiom). :grin:

File: funfact/algorithm.py [hunk 2]
```diff
@@ -113,22 +115,23 @@ class _Factorization(Factorization, ab.AutoGradMixin):
     opt_fac = _Factorization.from_tsrex(tsrex_vec, dtype=dtype)
 
     try:
-        opt = optimizer(opt_fac.factors, lr=lr, **kwargs)
+        opt = optimizer(opt_fac.factors, lr=lr)
```
Comment (campsd) [inline funfact/algorithm.py:118]: This means that we can not set the other hyperparameters of the optimizer? E.g. `beta1`, `beta2`, `epsilon` for Adam. Maybe we can use the dictionary approach to still allow this?

Comment (yhtang) [reply funfact/algorithm.py:118]: I am thinking about adapting the class-or-object idiom to optimizers too. So if the user supplies a class (or str of the class name), it will be instantiated with the default arguments. Alternatively, they can create an optimizer object on their own so that additional arguments can be given there.

File: funfact/test_loss.py [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+
+    loss = loss_cls()
+    for shape in [(3,), (3, 3), (3, 4), (3, 4, 5), (2, 3, 4, 5)]:
+        for _ in range(64):
+            a = ab.normal(0.0, 1.0, shape)
+            b = ab.normal(0.0, 1.0, shape)
+            assert loss(a, a) == pytest.approx(0, abs=1e-6)
+            assert loss(b, b) == pytest.approx(0, abs=1e-6)
+            assert loss(a, b) >= 0
+            assert loss(b, a) >= 0
```
Comment (campsd) [inline funfact/test_loss.py:26]: Maybe we can add a test that `loss(a, b)` and `loss(b,a)` should be approximately equal?

File: funfact/test_loss.py [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+    MSE,
+    L1,
+    KLDivergence
+])
+def test_reduction(loss_cls):
+
+    loss_sum = loss_cls(reduction='sum')
+    loss_mean = loss_cls(reduction='mean')
+    with pytest.raises(Exception):
+        loss_cls(reduction='abrakadabra')
```
Comment (campsd) [inline funfact/test_loss.py:109]: Great reduction method 😆

File: funfact/algorithm.py (hunk 2 above)
Comment (yhtang) [reply funfact/algorithm.py:118]: In other words, we should avoid bringing the keywords of components such as loss or optimizer into the factorize method, because:
1. It will become more and more challenging if we involve more components. For example, I am eyeing multi-tensor penalty terms for V2.0 :grin: 
2. This would make the signature more and more complicated, and also requires us to duplicate docstrings.

Comment (campsd) [reply funfact/algorithm.py:118]: Sounds like a good idea

File: funfact/test_loss.py (hunk 3 above)
Comment (yhtang) [reply funfact/test_loss.py:26]: Good point.

File: funfact/test_loss.py (hunk 4 above)
Comment (yhtang) [reply funfact/test_loss.py:109]: :sunglasses:

File: funfact/test_loss.py (hunk 3 above)
Comment (campsd) [reply funfact/test_loss.py:27]: Yes, but I believe this test only runs for Lp norms :)

Comment (yhtang) [reply funfact/test_loss.py:27]: Done

**Review by campsd — APPROVED** [review]

File: funfact/test_loss.py (hunk 3 above)
Comment (yhtang) [reply funfact/test_loss.py:27]: > Yes, but I believe this test only runs for Lp norms :)

Yeah, I realized it as soon as I opened up the file. Seems like I already forgot part of what I wrote two days ago. :joy:
